### PR TITLE
Add validating webhook for user and identity

### PR DESCRIPTION
**Pull request by wanghaoran1988**
https://issues.redhat.com/browse/OSD-3271

### CODE REVIEW

File: src/webhook/identity_validation.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    userinfo = body_dict['userInfo']
+    if userinfo['username'] in ("kube:admin", "system:admin", "system:serviceaccount:openshift-authentication:oauth-openshift"):
+      # kube/system admin can do anything
+      if debug:
+        print("Performing action: {} on identity {} by {}".format(body_dict['operation'], identity_name, userinfo['username']))
+      return responses.response_allow(req=body_dict)
+ 
+    sre_user = False
+    for id_prefix in identity_prefix_list:
+      if identity_name.startswith(id_prefix):
```
Comment (jewzaam) [inline src/webhook/identity_validation.py:57]: We don't create any identity other than `OpenShift_SRE`, do we need a prefix check or just equals?

File: src/webhook/user_validation.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+from prometheus_client import Counter
+from webhook.request_helper import validate, responses
+
+bp = Blueprint("user-webhook", __name__)
+
+# define what we track, declare Counter, how many times this route is accessed
+TOTAL_USER = Counter('webhook_user_validation_total', 'The total number of user validation requests')
+DENIED_USER = Counter('webhook_user_validation_denied', 'The total number of user validation requests denied')
+
+user_suffix = "@redhat.com"
```
Comment (jewzaam) [inline src/webhook/user_validation.py:13]: Recommend renaming `user_suffix` to something more descriptive like `protected_user_suffix`

File: src/webhook/identity_validation.py (hunk 1 above)
Comment (wanghaoran1988) [reply src/webhook/identity_validation.py:57]: @jew I thought we may add others for other teams like RHMI and CEE, if we don't, I will remove the list. Here I am using identity name to check the prefix, it has `OpenShift_SRE:Z3NsZWVtYW4` format, I think I should use `providerName` name, so a equals makes more sense.

**Comment by wanghaoran1988** [comment]
@jewzaam Unit test added, PTAL

File: src/webhook/test_identity_validation.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                    "uid": "testuser",
+                    "operation": "UPDATE",
+                    "userInfo": {
+                        "groups": groups,
+                        "username": username,
+                    },
+                    "object": {
+                        "kind": "Identity",
+                        "metadata": {
+                            "name": "identityName"
```
Comment (jewzaam) [inline src/webhook/test_identity_validation.py:25]:
```suggestion
                            "name": identityName
```

File: src/webhook/test_identity_validation.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        self.assertTrue(response['response']['allowed'])
+
+    # kube:admin can update redhat user identity
+    def test_kube_admin(self):
+        response = self.runtests(
+            "kube:admin", "", "OpenShift_SRE:test", "OpenShift_SRE")
+        self.assertTrue(response['response']['allowed'])
+
+    # users in sre admin groups can update redhat user identity
+    def test_sre_groups(self):
```
Comment (jewzaam) [inline src/webhook/test_identity_validation.py:66]: Recommend add a test to verify SRE can edit a non-SRE identity.

File: src/webhook/test_identity_validation.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+]
+
+
+def create_request(username, groups, identityName, providerName):
+    class FakeRequest(object):
+        def __init__(self):
+            self.json = {
+                "request": {
+                    "uid": "testuser",
+                    "operation": "UPDATE",
```
Comment (jewzaam) [inline src/webhook/test_identity_validation.py:17]: Should we write test to iterate over the operations we care about and just make sure we don't have regressions?  Could verify CREATE, UPDATE, and DELETE.

**Comment by jewzaam** [comment]
@wanghaoran1988 same comments for test_user_validation.py as I put on test_identity_validation.py

File: src/webhook/test_identity_validation.py (hunk 5 above)
Comment (wanghaoran1988) [reply src/webhook/test_identity_validation.py:17]: No, we can only do it in e2e tests, this is only unit test, the validation func doesn't check whether it's "CREATE" or "UPDATE", the verbs are configured in `ValidatingWebhookConfiguration`.

File: src/webhook/test_identity_validation.py (hunk 4 above)
Comment (wanghaoran1988) [reply src/webhook/test_identity_validation.py:66]: done

File: src/webhook/test_identity_validation.py (hunk 3 above)
Comment (wanghaoran1988) [reply src/webhook/test_identity_validation.py]: good catch

**Comment by wanghaoran1988** [comment]
@jewzaam Comments addressed, could you take another look?

**Pull request merged**